### PR TITLE
Paystation test

### PR DESCRIPTION
**Pull request by WBGroupR**
Please review PaystationControllerTest. Note that most tests fail due to Null Pointer Exception due to different barcode string format - to be fixed is all classes that use it.
Also, tests based on controller state cannot properly test state as this is implemented currently as a display string in user interface rather than a state.
For comment.

### CODE REVIEW

File: test/bcccp/carpark/CarparkTest.java [hunk 1]
```diff
@@ -23,7 +27,7 @@
 
 import static org.junit.jupiter.api.Assertions.*;
 
-@RunWith(MockitoJUnitRunner.class)
```
Comment (duane-mcmahon) [inline test/bcccp/carpark/CarparkTest.java:26]: you have made some minor changes to the carpark test class. was mockito not working for you?
you may not have intended to commit changes to this class.

**Review by duane-mcmahon — APPROVED** [review]
OK! so it appears that you have submitted valid Junit tests for the paystation test class, incorporating verify() and ArgumentCaptor from the mockito mocking framework. When reviewing tests I guess - since we have yet to really discuss the evaluation criteria for writing tests - that i am looking to determine whether:
-I understand the tests?
-they test non-trivial functionality?
-they test important functionality?
                        -they test important edge cases?
                        -they reasonably cover success cases?
                        -they reasonably cover failure cases?
So i am checking to see that the tests at least cover the confusing or complicated sections of code in the module/class being submitted.

Ok, so it would appear that you have tested for most possibilities if not all of the important ones. In relation to the BCCCP Software Specification:

The constructor has not been tested to demonstrate that RuntimeException is thrown if one of the arguments supplied to the constructor is null.

Several of your tests throw nullpointerexceptions, should tests throw exceptions??  

In the @BeforeEach you create a barcode using a method that itself will require independent testing (i'll do it!).

i am happy to approve these tests provisionally because they will contribute to a feedback loop and the evolution of our codebase.

Thx for the work,

Duane.

**Comment by duane-mcmahon** [comment]
I merge this into master at 11 pm tonight if there is no further discussion by then.

**Comment by WBGroupR** [comment]
Thanks Duane,

Yes, your comments are valid. I found it very difficult to test these methods as they were void and some also had no arguments. But Mockito allows you to at least verify if those void methods called other methods. The issue with the controller classes is that they are supposed to check state, but the “state” is not implemented as a true state, is instead just passing a string to be displayed in the user interface - so I try to test if the method is at least passing the correct string to the mocked user interface display() method.

True, it would be missing the test on constructor.

Regarding nullpointerexceptions, the tests that throw them are the ones that refer to adHocTicket methods in mocked carpark class probably I think because there is a difference in the barcode format that I pass as opposed to the one expected, so no ticket is found (null). So, yes, the barcode format needs to be standardised across all modules.

Cheers,
Wayne

File: test/bcccp/carpark/CarparkTest.java (hunk 1 above)
Comment (WBGroupR) [reply test/bcccp/carpark/CarparkTest.java:26]: No changes made to carpark - at least not intentionally. That is a worry. I may need to re-fetch and pull carpark to my local master and then re-merge with my local branch, etc. Also, I have noticed that CarparkTest includes some strange corruptions in the code, see lines 65, 68 and 72 - these look like git commands/output have somehow been merged into the code - not sure how that could have happened.
